drivers/sdhc/sdhci_common.c

```diff
@@ -0,0 +1,794 @@
+{
+	uint16_t ctrl;
+
+	ctrl = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+	switch (signal_voltage) {
+	case SD_VOL_3_3_V:
+		ctrl &= ~SDHCI_CTRL_VDD_180;
+		sys_write16(ctrl, sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+		/* Wait for 5ms */
+		k_msleep(5);
+
+		/* 3.3V regulator output should be stable within 5 ms */
+		ctrl = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+		LOG_DBG("sdhc:set_3_3: SDHCI_HOST_CONTROL2: %04x", ctrl);
+		if (ctrl & SDHCI_CTRL_VDD_180) {
+			LOG_ERR("sdhc: 3.3V regulator output did not become stable\n");
+			return -EIO;
+		}
+
+		break;
+	case SD_VOL_1_8_V:
+		ctrl |= SDHCI_CTRL_VDD_180;
+		sys_write16(ctrl, sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+		/* Wait for 5 ms */
+		k_msleep(5);
+
+		/* 1.8V regulator output has to be stable within 5 ms */
+		ctrl = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+		LOG_DBG("sdhc:set_1_8: SDHCI_HOST_CONTROL2: %04x", ctrl);
+		if (!(ctrl & SDHCI_CTRL_VDD_180)) {
+			LOG_ERR("sdhc: 1.8V regulator output did not become stable\n");
+			return -EIO;
+		}
+
+		break;
+	default:
+		LOG_DBG("sdhc: signal voltage not supported %d", signal_voltage);
+		return -ENOTSUP;
+	}
+
+	return 0;
+}
+
+int sdhci_set_uhs_timing(struct sdhci_common *sdhci_ctx, enum sdhc_timing_mode timing)
+{
+	bool hspd = false;
+	uint16_t reg;
+
+	reg = sys_read16(sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+	reg &= ~SDHCI_CTRL_UHS_MASK;
+
+	switch (timing) {
+	case SDHC_TIMING_LEGACY:
+	case SDHC_TIMING_SDR12:
+		reg |= SDHCI_CTRL_UHS_SDR12;
+		break;
+	case SDHC_TIMING_SDR25:
+		reg |= SDHCI_CTRL_UHS_SDR25;
+		hspd = true;
+		break;
+	case SDHC_TIMING_SDR50:
+		reg |= SDHCI_CTRL_UHS_SDR50;
+		break;
+	case SDHC_TIMING_DDR50:
+		reg |= SDHCI_CTRL_UHS_DDR50;
+		break;
+	case SDHC_TIMING_SDR104:
+		reg |= SDHCI_CTRL_UHS_SDR104;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	sdhci_enable_sd_clock(sdhci_ctx, false);
+
+	LOG_DBG("sdhc:timing: SDHCI_HOST_CONTROL2: %04x", reg);
+	sys_write16(reg, sdhci_ctx->reg_base + SDHCI_HOST_CONTROL2);
+
+	sdhci_set_hspd(sdhci_ctx, hspd);
+
+	sdhci_enable_sd_clock(sdhci_ctx, true);
+	return 0;
+}
+
+static int sdhc_stop_transfer(struct sdhci_common *sdhci_ctx)
+{
+	struct sdhc_command hdc_cmd = {0};
+
+	hdc_cmd.opcode = SD_STOP_TRANSMISSION;
+	hdc_cmd.arg = 0;
+	hdc_cmd.response_type = SD_RSP_TYPE_R1;
+	hdc_cmd.timeout_ms = 1000;
+
+	return sdhci_send_cmd(sdhci_ctx, &hdc_cmd, false);
+}
+
+static bool sdhc_need_cmd12(struct sdhc_command *cmd)
+{
+	return (cmd->opcode == SD_READ_MULTIPLE_BLOCK || cmd->opcode == SD_WRITE_MULTIPLE_BLOCK);
+}
+
+static int sdhci_poll_data_complete(struct sdhci_common *sdhci_ctx, int32_t timeout)
+{
+	uint32_t reg_normal_int_stat = 0;
+	int ret = -ETIMEDOUT;
+
+	timeout = timeout ? timeout : 1;
+	timeout = timeout * 100;
+
+	while (timeout > 0) {
+		reg_normal_int_stat = sys_read32(sdhci_ctx->reg_base + SDHCI_INT_STATUS);
+		if (reg_normal_int_stat & SDHCI_INT_XFER_COMPLETE) {
```

Comment:
This function is called for both DMA and non-DMA modes. Should we check another flag here, such as `SDHCI_INT_DMA_END`, or even `SDHCI_INT_DATA_MASK`, in case an error occurs earlier than the timeout expiration?

Reply:
No. No DMA events is expected here - DMA is a specific event which is not used by current implementation.
Error checked below, but yeah may be checked in loop also.